core/templates/hash_map2.h

```diff
@@ -0,0 +1,539 @@
+	};
+
+	struct ConstIterator {
+		_FORCE_INLINE_ const KeyValue<TKey, TValue> &operator*() const {
+			return E->data;
+		}
+		_FORCE_INLINE_ const KeyValue<TKey, TValue> *operator->() const { return &E->data; }
+		_FORCE_INLINE_ ConstIterator &operator++() {
+			if (E) {
+				E = E->next;
+			}
+			return *this;
+		}
+		_FORCE_INLINE_ ConstIterator &operator--() {
+			if (E) {
+				E = E->prev;
+			}
+			return *this;
+		}
+
+		_FORCE_INLINE_ bool operator==(const ConstIterator &b) const { return E == b.E; }
+		_FORCE_INLINE_ bool operator!=(const ConstIterator &b) const { return E != b.E; }
+
+		ConstIterator(const HashMap2Element<TKey, TValue> *p_E) { E = p_E; }
+		ConstIterator() {}
+		ConstIterator(const ConstIterator &p_it) { E = p_it.E; }
+
+	private:
+		const HashMap2Element<TKey, TValue> *E = nullptr;
+	};
+
+	_FORCE_INLINE_ Iterator begin() {
+		return Iterator(head_element);
+	}
+	_FORCE_INLINE_ Iterator end() {
+		return Iterator(nullptr);
+	}
+	_FORCE_INLINE_ Iterator last() {
+		return Iterator(tail_element);
+	}
+
+	_FORCE_INLINE_ Iterator find(const TKey &p_key) {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return end();
+		}
+		return Iterator(elements[pos]);
+	}
+
+	_FORCE_INLINE_ void remove(const Iterator &p_iter) {
+		if (p_iter.E) {
+			erase(p_iter.E->data.key);
+		}
+	}
+
+	_FORCE_INLINE_ ConstIterator begin() const {
+		return ConstIterator(head_element);
+	}
+	_FORCE_INLINE_ ConstIterator end() const {
+		return ConstIterator(nullptr);
+	}
+	_FORCE_INLINE_ ConstIterator last() const {
+		return ConstIterator(tail_element);
+	}
+
+	_FORCE_INLINE_ ConstIterator find(const TKey &p_key) const {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return end();
+		}
+		return ConstIterator(elements[pos]);
+	}
+
+	/* Indexing */
+
+	const TValue &operator[](const TKey &p_key) const {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		CRASH_COND(!exists);
+		return elements[pos]->data.value;
+	}
+
+	TValue &operator[](const TKey &p_key) {
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return _insert(p_key, TValue())->data.value;
+		} else {
+			return elements[pos]->data.value;
+		}
+	}
+
+	/* Constructors */
+
+	HashMap2(const HashMap2 &p_other) {
+		(*this) = p_other;
+	}
+
+	void operator=(const HashMap2 &p_other) {
+		if (this == &p_other) {
+			return; // Ignore self assignment.
+		}
+		if (capacity != 0) {
+			clear();
+		}
+
+		reserve(p_other.capacity);
+
+		if (p_other.elements == nullptr) {
+			return; // Nothing to copy.
+		}
+
+		for (uint32_t i = 0; i < p_other.capacity; i++) {
+			if (p_other.hashes[i] == EMPTY_HASH) {
+				continue;
+			}
+
+			insert(p_other.elements[i]->data.key, p_other.elements[i]->data.value);
+		}
+	}
+
+	HashMap2(uint32_t p_initial_capacity = 64) {
+		// Capacity can't be 0.
+		capacity = MAX(MIN_CAPACITY, p_initial_capacity);
```

Comment:
```suggestion
		capacity = MAX((uint32_t)MIN_CAPACITY, p_initial_capacity);
```

GCC seems to hate it:
```
./core/typedefs.h:122:19: error: enumeral and non-enumeral type in conditional expression [-Werror=extra]
  122 |  return m_a > m_b ? m_a : m_b;
      |         ~~~~~~~~~~^~~~~~~~~~~
./core/typedefs.h: In instantiation of 'constexpr auto MAX(T, T2) [with T = HashMap2<Vector2i, Vector2>::<unnamed enum>; T2 = unsigned int]':
./core/templates/hash_map2.h:525:17:   required from 'HashMap2<TKey, TValue, Hasher, Comparator, Allocator>::HashMap2(uint32_t) [with TKey = Vector2i; TValue = Vector2; Hasher = HashMapHasherDefault; Comparator = HashMapComparatorDefault<Vector2i>; Allocator = DefaultTypedAllocator<HashMap2Element<Vector2i, Vector2> >; uint32_t = unsigned int]'
```

Reply:
The comment about "Capacity can't be 0" is also outdated. It was for using 1 as min capacity.